### PR TITLE
Fixed #22106 -- Make django.views.i18n.javascript_catalog more flexible

### PR DESCRIPTION
**Pull request by MoritzS**
See https://code.djangoproject.com/ticket/22106 for more information

### CODE REVIEW

**Comment by timgraham** [comment]
I know very little about `javascript_catalog`, but maybe if you could rebase your patch and add some documentation and release notes as described in our [patch review checklist](https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/submitting-patches/#patch-review-checklist), that will help me.

**Comment by MoritzS** [comment]
I rebased it and added some documentation.

**Comment by timgraham** [comment]
buildbot, test this please.

**Comment by timgraham** [comment]
buildbot, test on selenium.

**Comment by timgraham** [comment]
Test failure on Python 2: 
```
Traceback (most recent call last):
  File "/mnt/jenkinsdata/workspace/pull-requests-selenium/database/sqlite3/python/python2.7/django/test/utils.py", line 182, in inner
    return test_func(*args, **kwargs)
  File "/mnt/jenkinsdata/workspace/pull-requests-selenium/database/sqlite3/python/python2.7/django/test/utils.py", line 182, in inner
    return test_func(*args, **kwargs)
  File "/mnt/jenkinsdata/workspace/pull-requests-selenium/database/sqlite3/python/python2.7/tests/view_tests/tests/test_i18n.py", line 279, in test_multiple_catalogs
    self.assertEqual(elem.text, "il faut traduire cette chaîne de caractères de app1")
AssertionError: u'il faut traduire cette cha\xeene de caract\xe8res de app1' != 'il faut traduire cette cha\xc3\xaene de caract\xc3\xa8res de app1'
```

File: django/views/i18n.py [hunk 1]
```diff
@@ -94,87 +94,87 @@ def get_formats():
...
-
-  django.formats = {{ formats_str }};
-
-  django.get_format = function (format_type) {
-    var value = django.formats[format_type];
-    if (typeof(value) == 'undefined') {
-      return format_type;
-    } else {
+  if (!django.jsi18n_initialized) {
+    django.gettext = function (msgid) {
```
Comment (timgraham) [inline django/views/i18n.py]: I think we should drop the space after "function" through this file (can be a separate commit).

File: docs/releases/1.9.txt [hunk 2]
```diff
@@ -130,6 +130,9 @@ Internationalization
 * The :func:`django.views.i18n.set_language` view now properly redirects to
   :ref:`translated URLs <url-internationalization>`, when available.
 
+* The :func:`django.views.i18n.javascript_catalog` view now correctly works
```
Comment (timgraham) [inline docs/releases/1.9.txt]: works correctly

File: docs/releases/1.9.txt [hunk 3]
```diff
@@ -130,6 +130,9 @@ Internationalization
...
+  if used multiple times with different configuration on the same page.
```
Comment (timgraham) [inline docs/releases/1.9.txt]: configurations

File: docs/topics/i18n/translation.txt [hunk 4]
```diff
@@ -943,6 +943,31 @@ different apps and this changes often and you don't want to pull in one big
 catalog file. As a security measure, these values can only be either
 ``django.conf`` or any package from the :setting:`INSTALLED_APPS` setting.
 
+You can also split the catalogs in multiple urls and load them as you need in
```
Comment (timgraham) [inline docs/topics/i18n/translation.txt]: URLs

File: docs/topics/i18n/translation.txt [hunk 5]
```diff
@@ -943,6 +943,31 @@ different apps and this changes often and you don't want to pull in one big
...
+    js_info_dict_app = {
+        'packages': ('your.app.package',),
+    }
+
+    js_info_dict_other_app = {
+        'packages': ('your.other.app.package',),
+    }
+
+    urlpatterns = [
+        url(r'^jsi18n/app$', javascript_catalog, js_info_dict_app),
```
Comment (timgraham) [inline docs/topics/i18n/translation.txt]: `/$` (most URL examples use a trailing slash)

File: docs/topics/i18n/translation.txt [hunk 6]
```diff
@@ -943,6 +943,31 @@ different apps and this changes often and you don't want to pull in one big
...
+        url(r'^jsi18n/other_app$', javascript_catalog, js_info_dict_other_app),
+    ]
+
+If you use more than one ``javascript_catalog`` on a site and some of them
+define the same strings, the ones in the catalog that was loaded latest
```
Comment (timgraham) [inline docs/topics/i18n/translation.txt]: ones -> strings
latest -> last

File: docs/topics/i18n/translation.txt [hunk 7]
```diff
@@ -943,6 +943,31 @@ different apps and this changes often and you don't want to pull in one big
...
+take precedence.
+
+.. versionchanged:: 1.9
+
+Before Django 1.9, the catalogs completely overwrote each other and you could
```
Comment (timgraham) [inline docs/topics/i18n/translation.txt]: indent 4 spaces

**Comment by timgraham** [comment]
Please bump the ticket to "ready for checkin" after making those updates, thanks!

**Comment by timgraham** [comment]
buildbot, test on selenium.

**Comment by MoritzS** [comment]
There seems to be a problem with the build bot, the errors come from Java Exceptions.

**Comment by timgraham** [comment]
buildbot, test on selenium.

File: tests/view_tests/tests/test_i18n.py [hunk 8]
```diff
@@ -270,6 +270,16 @@ def test_javascript_gettext(self):
         elem = self.selenium.find_element_by_id("npgettext_plur")
         self.assertEqual(elem.text, "455 Resultate")
 
+    @modify_settings(INSTALLED_APPS={'append': ['view_tests.app1', 'view_tests.app2']})
+    @override_settings(LANGUAGE_CODE='fr')
+    def test_multiple_catalogs(self):
+        self.selenium.get('%s%s' % (self.live_server_url, '/jsi18n_multi_catalogs/'))
+
+        elem = self.selenium.find_element_by_id('app1string')
+        self.assertEqual(elem.text.encode('utf8'), b'il faut traduire cette cha\xc3\xaene de caract\xc3\xa8res de app1')
```
Comment (claudep) [inline tests/view_tests/tests/test_i18n.py]: Why do you encode then test the bytestring, instead of testing the original unicode result?

Comment (MoritzS) [inline tests/view_tests/tests/test_i18n.py]: I think that's the most portable way for testing this with python 2 and 3.
Another way would be to test:
```python
self.assertEqual(elem.text, u'il faut traduire cette chaîne de caractères de app1')
```
But I think, it's more portable not depending on how selenium or the current python version interpret the string.

Comment (timgraham) [inline tests/view_tests/tests/test_i18n.py]: You can add `from __future__ import unicode_literals` to that test file.

Comment (MoritzS) [inline tests/view_tests/tests/test_i18n.py]: I was not sure if that's ok.
So I'll just do that now.

**Comment by timgraham** [comment]
merged in 6bb2175ed65456dda2f8ba584f206ba4bbd43dbc, thanks!
